lib/exceptional_synchrony/faraday_monkey_patch.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+# Monkey patch for the Faraday method that creates the adapter used for a connection.
+# If the thread local variable :running_em_synchrony is true, it overrides this method
+# in order to force use of the :em_synchrony adapter rather than the :net_http adapter.
+# This ensures that the Eventmachine reactor does not get blocked by connection i/o.
+# This patch was built against faraday v0.17 and v1.3, although for v1.3 ruby2_keywords
+# usage was dropped in adapter method definition to simplify this code.
+begin
+  require 'faraday'
+
+  module Faraday
+    class RackBuilder
+      def adapter(klass = NO_ARGUMENT, *args, &block)
+        if Faraday::VERSION[0] == "0"
+
+          # BEGIN PATCH
+          key = klass
+          if key == :net_http && Thread.current.thread_variable_get(:running_em_synchrony)
+            key = :em_synchrony
+          end
+          # END PATCH
+
+          use_symbol(Faraday::Adapter, key, *args, &block)
+        else
+          return @adapter if klass == NO_ARGUMENT
+
+          klass = Faraday::Adapter.lookup_middleware(klass) if klass.is_a?(Symbol)
+
+          # BEGIN PATCH
+          if klass == Faraday::Adapter::NetHttp && Thread.current.thread_variable_get(:running_em_synchrony)
+            klass = Faraday::Adapter::EMSynchrony
+          end
+          # END PATCH
+
+          @adapter = self.class::Handler.new(klass, *args, &block)
+        end
+      end
+    end
+  end
```

Comment:
Can you do this with a module that's prepended in front of `Faraday::RackBuilder`? In fact 2 modules, one for each of the Faraday versions would be really clear, because that will put the version check outside and keep the modules clear, and make it obvious which one we can throw away when we've stopped supporting that version.

Reply:
I don't understand. Are you suggesting something like
```module Faraday_v0
  module Faraday
    class RackBuilder
      ... v0 patch
    end
  end
end```

But that would hide the patch, so that must not be what you mean. How would those 2 modules be selectively applied?

Reply:
Similar, but like this:
```
begin
  require 'faraday'

  module ExceptionalSynchrony
    module FaradayAdapterPatch_v0
      def adapter(klass = NO_ARGUMENT, *args, &block)
        ...
      end
    end

    module FaradayPatch_v1
      def adapter(klass = NO_ARGUMENT, *args, &block)
        ...
      end
    end
  end

  if Faraday::VERSION.start_with(?"0")
    Faraday::RackBuilder.prepend ExceptionalSynchrony::FaradayAdapterPatch_v0
  else
    Faraday::RackBuilder.prepend ExceptionalSynchrony::FaradayAdapterPatch_v1
  end
rescue LoadError
  # patch is not needed if faraday is not available
end
```

Reply:
Or, this is equivalent I think:
```
begin
  require 'faraday'

  module ExceptionalSynchrony
    if Faraday::VERSION.start_with(?"0")
      module FaradayAdapterPatch
        def adapter(klass = NO_ARGUMENT, *args, &block)
          ...
        end
      end
    else
      module FaradayAdapterPatch
        def adapter(klass = NO_ARGUMENT, *args, &block)
          ...
        end
      end
    end
  end

  Faraday::RackBuilder.prepend ExceptionalSynchrony::FaradayAdapterPatch

rescue LoadError
  # patch is not needed if faraday is not available
end
```